packages/webamp/js/actionCreators/files.ts

```diff
@@ -400,3 +401,62 @@ export function addDirAtIndex(nextIndex: number): Thunk {
     );
   };
 }
+
+export function addFilesFromUrl(atIndex = 0): Thunk {
+  return async (dispatch, getState, { handleAddUrlEvent }) => {
+    if (handleAddUrlEvent) {
+      const tracks = await handleAddUrlEvent();
+
+      if (tracks != null) {
+        dispatch(loadMediaFiles(tracks, LOAD_STYLE.NONE, atIndex));
+        return;
+      }
+    } else {
+      alert("Not supported in Webamp");
+    }
+  };
+}
+
+export function addFilesFromList(): Thunk {
+  return async (dispatch, getState, { handleLoadListEvent }) => {
+    if (handleLoadListEvent) {
+      const tracks = await handleLoadListEvent();
+
+      if (tracks != null) {
+        dispatch(removeAllTracks());
+
+        dispatch(loadMediaFiles(tracks, LOAD_STYLE.NONE, 0));
+        return;
+      }
+    } else {
+      alert("Not supported in Webamp");
+    }
+  };
+}
+
+export function saveFilesToList(): Thunk {
+  return (dispatch, getState, { handleSaveListEvent }) => {
+    if (handleSaveListEvent) {
+      const state = getState();
+      const trackIds = getTrackOrder(state);
+      const tracks = getTracks(state);
+      handleSaveListEvent(
+        trackIds.map((id) => {
+          const track = tracks[id];
+
+          return {
+            url: track.url,
+            metaData: {
+              artist: track.artist || "",
```

Comment:
I'm generally not a fan of `|| ""`. If the field is nullable, pass null, if it's not nullable then it's required and passing it an empty value is going to cause unexpected behavior. Could we just null out the entire `metaData` object if any of the required fields are missing?

Reply:
I try to follow same paths than here: https://github.com/captbaritone/webamp/blob/da7e20c0f3de3460cf9f93dca7101fe4a0e68fac/packages/webamp/js/actionCreators/files.ts#L318

Reply:
Touché 😛